### PR TITLE
Max Over Time in imdb_cnn.py

### PR DESCRIPTION
**Pull request by dbonadiman**
Following this issue https://github.com/fchollet/keras/issues/2296 i propose this PR.
The mayor optimisation a part of the Max over time are:

- Dropout in the Embedding layer.
- Longer input sequences (400 instead of 100), made possible from the speedup of the Max Over Time.
- Adam optimizer.

Overall it takes 90 to 100 sec per epoch on my laptop CPU and in two epochs it reaches 0.885 accuracy that is a 5 points improvement over the previous implementation. Moreover it requires less memory (300k parameters vs 3M+) since the number of parameters do not depend  by the length of the input sequence anymore.

### CODE REVIEW

**Comment by dbonadiman** [comment]
The benchmark are all done using theano, it works even on tf by the way (with similar results) but it requires much more time and memory due to some "Embedding layer" related issue. 

`(Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.)`

However, this warning is not related with the pull request i get it every time.

**Comment by gaa-cifasis** [comment]
You have not tried with a GPU?

**Comment by dbonadiman** [comment]
I don't have Tensorflow on the server however with Theano on a Tesla K40:
Epoch 1/2
20000/20000 [==============================] - 10s - loss: 0.4580 - acc: 0.7641 - val_loss: 0.3159 - val_acc: 0.8638
Epoch 2/2
20000/20000 [==============================] - 10s - loss: 0.2937 - acc: 0.8773 - val_loss: 0.2929 - val_acc: 0.8778

**Comment by gaa-cifasis** [comment]
10s per iteration is great!  You should probably report that in the header of imdb_cnn.py.

**Comment by dbonadiman** [comment]
Actually, when working on GPU you can do even better both in terms of time and performances (More epochs, greater batches, highest dropout, convolutional filter of size 5, optionally zero padding for the borders of the sequence and even longer sequences likes don't cut any sequence -> size 2821).

But i think this is out of the scope of a CNN tutorial example.

Epoch 1/10
20000/20000 [==============================] - 46s - loss: 0.5322 - acc: 0.7043 - val_loss: 0.3307 - val_acc: 0.8732
Epoch 2/10
20000/20000 [==============================] - 46s - loss: 0.3146 - acc: 0.8649 - val_loss: 0.2710 - val_acc: 0.9008
Epoch 3/10
20000/20000 [==============================] - 46s - loss: 0.2325 - acc: 0.9076 - val_loss: 0.2531 - val_acc: 0.8972
Epoch 4/10
20000/20000 [==============================] - 46s - loss: 0.1733 - acc: 0.9337 - val_loss: 0.2440 - val_acc: 0.9004
Epoch 5/10
20000/20000 [==============================] - 46s - loss: 0.1395 - acc: 0.9475 - val_loss: 0.2408 - val_acc: 0.9010

**Comment by gaa-cifasis** [comment]
> Actually, when working on GPU you can do even better both in terms of time and performances (More epochs, greater batches, highest dropout, convolutional filter of size 5, optionally zero padding for the borders of the sequence and even longer sequences likes don't cut any sequence -> size 2821).

Neat!

> But i think this is out of the scope of a CNN tutorial example.

You are probably right about this. Still, some people can be interested in convolutional models that are better than using a svm with a bow (as well as some hand-crafted features), so i'm saving this discussion as a reference :smile:

**Comment by tboquet** [comment]
On a GTX980ti, with TensorFlow:
```python
Train on 20000 samples, validate on 5000 samples
Epoch 1/2
20000/20000 [==============================] - 17s - loss: 0.4585 - acc: 0.7630 - val_loss: 0.3045 - val_acc: 0.8754
Epoch 2/2
20000/20000 [==============================] - 16s - loss: 0.2896 - acc: 0.8783 - val_loss: 0.3048 - val_acc: 0.8702
```

**Comment by fchollet** [comment]
Sounds good. Your max function is just an extreme version of MaxPooling1D though (which also works over time). Do you have a justification as to why it would work better? Intuitively I would think the main change here and the reason you're seeing better performance is because you're using much longer sequences, carrying much more information.

**Comment by dbonadiman** [comment]
I agree with you the main reason to add the Lambda is that the same version with MaxPooling1D is the following.

`MaxPooling1D(pool_lenght=model.layers[-1 ].output_shape[1]) `

That looks a bit uglier.

Using a standard Max is nicer since it allows to remove the Flatten Layer too.

However i agree the mayor reason is that you can process longer sequences (and theoretically of variable size) without affecting the number of parameter.

**Comment by fchollet** [comment]
Looks good to me.

Why are you using Adam? Does it work better than RMSProp? Note that Adam will run somewhat slower than RMSProp.

Please add runtime on GPU in the docstring, since you have it.

**Comment by dbonadiman** [comment]
Adam gives 1 point improvement there, probably task related with the default parameter, i will add this value soon, thanks.

**Comment by dbonadiman** [comment]
@fchollet Maybe a comment should be added in the code indicating that one may use K.mean or K.sum without changing anything else to experiment with other types Pooling. It results in similar results.

**Comment by ekerazha** [comment]
@dbonadiman 
Using the Lambda layer instead of MaxPooling1D doesn't work anymore since keras 1.0.5. The example is now broken.
